### PR TITLE
Added facebook auth and figaro. Uploaded facebook keys to heroku. Fac…

### PR DESCRIPTION
**Pull request by Mikeysax**
…ebook auth is set up to work on live server not development

### CODE REVIEW

**Comment by Mikeysax** [comment]
@lasley Think this is good.

**Comment by Mikeysax** [comment]
I'm not 100% on the routes.rb. I think I wrote it correctly? Or it might be:   match '/auth/:provider/callback' => 'authentications#create', :via => [:get, :post], :as => 'authentications_create'

**Comment by Mikeysax** [comment]
Or this:   get '/auth/:provider/callback', :to => 'authentications#create'

File: config/routes.rb [hunk 1]
```diff
@@ -1,4 +1,5 @@
 Rails.application.routes.draw do
+  get '/auth/:provider/callback' => 'authentications#create'
```
Comment (lasley) [inline config/routes.rb]: Should also add auth failure and session destroy routes. Take a look at this example - https://richonrails.com/articles/facebook-authentication-in-ruby-on-rails

File: .rubocop.yml [hunk 2]
```diff
@@ -8,3 +8,4 @@ AllCops:
   Exclude:
     - 'db/**/*'
     - 'bin/**/*'
+    - '**/authentications_controller.rb'
```
Comment (lasley) [inline .rubocop.yml]: Why exclude?

File: app/views/devise/registrations/new.html.erb [hunk 3]
```diff
@@ -15,7 +24,11 @@
 
       <div class="form-actions">
         <%= f.button :submit, "Sign up" %>
+        <br />
+        <!-- <div class="facebook"><%= link_to 'Login with Facebook', '/auth/facebook/' %></div> -->
```
Comment (lasley) [inline app/views/devise/registrations/new.html.erb]: Remove old code

File: app/views/devise/sessions/new.html.erb [hunk 4]
```diff
@@ -12,9 +20,13 @@
 
       <div class="form-actions">
         <%= f.button :submit, "Log in" %>
+        <br />
+        <!-- <div class="facebook"><%= link_to 'Login with Facebook', '/auth/facebook/' %></div> -->
```
Comment (lasley) [inline app/views/devise/sessions/new.html.erb]: Remove old code

File: app/views/devise/sessions/new.html.erb [hunk 5]
```diff
@@ -12,9 +20,13 @@
 
       <div class="form-actions">
         <%= f.button :submit, "Log in" %>
```
Comment (lasley) [inline app/views/devise/sessions/new.html.erb]: `Log in` seems wrong with `Sign In/Up` being used everywhere else IMO

File: app/views/devise/sessions/new.html.erb [hunk 6]
```diff
@@ -1,7 +1,15 @@
+<div id="fb-root"></div>
+<script>(function(d, s, id) {
+  var js, fjs = d.getElementsByTagName(s)[0];
+  if (d.getElementById(id)) return;
+  js = d.createElement(s); js.id = id;
+  js.src = "//connect.facebook.net/en_US/sdk.js#xfbml=1&version=v2.5&appId=186146031741513";
+  fjs.parentNode.insertBefore(js, fjs);
+}(document, 'script', 'facebook-jssdk'));</script>
```
Comment (lasley) [inline app/views/devise/sessions/new.html.erb]: You should never use inline javascript. Take a look at http://theflyingdeveloper.com/controller-specific-assets-with-rails-4/

File: app/views/layouts/application.html.erb [hunk 7]
```diff
@@ -26,7 +26,24 @@
...
+    };
 
+    (function(d, s, id){
+       var js, fjs = d.getElementsByTagName(s)[0];
+       if (d.getElementById(id)) {return;}
+       js = d.createElement(s); js.id = id;
+       js.src = "//connect.facebook.net/en_US/sdk.js";
+       fjs.parentNode.insertBefore(js, fjs);
+     }(document, 'script', 'facebook-jssdk'));
+  </script>
```
Comment (lasley) [inline app/views/layouts/application.html.erb]: Should break this JS into an isolated file

File: app/views/layouts/application.html.erb [hunk 8]
```diff
@@ -42,6 +59,7 @@
                 <li class="nav-link-right"><%= link_to 'Sign In', new_user_session_path %></li>
                 <li class="nav-link-right"><%= link_to 'Sign Up', new_user_registration_path %></li>
               </div>
+              <!-- <li class="fb-like" data-share="true" data-width="450" data-show-faces="true"></li> -->
```
Comment (lasley) [inline app/views/layouts/application.html.erb]: Remove old code

File: app/models/user.rb [hunk 9]
```diff
@@ -6,4 +6,12 @@ class User < ActiveRecord::Base
 
   belongs_to :game
   has_many :chess_pieces
+  has_many :authentications, dependent: :delete_all
+
+  def apply_omniauth(auth)
```
Comment (lasley) [inline app/models/user.rb]: This should have an accompanying test. Take a look at https://github.com/intridea/omniauth/wiki/Integration-Testing

**Comment by lasley** [comment]
GJ on this - omniauth is a hard one.

**Comment by Mikeysax** [comment]
@lasley Should I start over or refactor?

File: app/views/devise/sessions/new.html.erb (hunk 5 above)
Comment (Mikeysax) [inline app/views/devise/sessions/new.html.erb]: Yeah, I didn't catch that one! Thank you!

File: .rubocop.yml (hunk 2 above)
Comment (Mikeysax) [inline .rubocop.yml]: It's giving me this: Assignment Branch Condition size for create is too high.  Doesn't that mean my method does too much or? I wasn't sure how to exactly fix it, unless I break it up into separate things?

**Comment by Mikeysax** [comment]
@lasley It is very hard, thank you! I'm learning a lot about omniauth, currently adding oauth flow to another application to have users create and store their stripe info for a marketplace.

**Comment by Mikeysax** [comment]
@lasley I'm running into a error where I'm trying to save the user to the database and It's giving me a validation error because I'm not taking in a email and password. Is there a way to make an exception for password as I've added email to be taken from the facebook account?

**Comment by Mikeysax** [comment]
@lasley Nvm I've figured it out.

File: .rubocop.yml (hunk 2 above)
Comment (lasley) [inline .rubocop.yml]: Not necessarily that you're doing too much, but that there are variable assignments nested deep in logic. This is typically a pain point in terms of long terms maintenance and requires refactoring (particularly if you're getting nailed on it in a controller). Take a look at http://codereview.stackexchange.com/questions/86405/reduce-assignment-branch-condition-size

If you do decide that you want to ignore the error, you should instead disable only that specific standard in the file - http://stackoverflow.com/questions/30174669/how-to-silence-rubocop-warning-on-assignment-branch-condition

**Comment by lasley** [comment]
Hahaha rubber ducky development at its finest above this comment :stuck_out_tongue: 

http://blog.codinghorror.com/rubber-duck-problem-solving/

**Comment by Mikeysax** [comment]
@lasley Ahhh that's so true, hah. Good read!

**Comment by Mikeysax** [comment]
@lasley I think this is good, I tried to rebase but I don't see the changes?. I've fixed the current_user = nil so now both facebook auth and devise auth work together properly.

**Comment by Mikeysax** [comment]
Actually, since I've uploaded I'm getting errors...

**Comment by Mikeysax** [comment]
application.yml file was deleted when I fixed a merge conflict but I readded it and cannot upload it to github.

**Comment by Mikeysax** [comment]
@lasley Don't know if the missing application.yml file will affect facebook login access for the rest of the team?

**Comment by Mikeysax** [comment]
@lasley Other than that everything is working as expected.

**Comment by lasley** [comment]
@Mikeysax :

Application.yml? You mean application.rb?

Also, what were you trying to rebase? Squashing commits? Make sure to `git push -f` it, or it will just crap out on you.

**Comment by lasley** [comment]
`-f` indicates a forceful operation. In simple terms, for a push, that means you're altering history. 

Any time you use `-f`, you are performing an operation that can cause issues for others that are working on the same branch.

**Pull request closed without merging**